### PR TITLE
Record layout

### PR DESCRIPTION
**Pull request by TwoClocks**
Here is my first cut at record layout ported from this Rust code https://github.com/mahkoh/repr-c.

I'm looking for feedback and/or a punch list before merging.

It's a bit delicate. If you look at it sideways one of the tests will likely break. I'm skipping a good % of the test cases for various reasons (#300, #301), as well as others I haven't dug into yet. There are also some tests that clang-13 fails on, so there is some discrepancy in Rust tests and clang. I need to look into those.

The tests are auto-generated from the Rust code. Because of that `test/test_records.zig` is doing the filtering for which tests to skip (or parts of tests to skip).

I shoved a bunch of stuff into `Type.Record` and `Type.Record.Field`. I think I can get rid of some of that in the future, but I'd like to leave it now as it makes debugging easier.

I haven't done any work towards getting other targets to work. I'm currently just trying to mimic clang on x86 linux.

There are some `asserts` as well as `panics`, and I haven't done any errors or warnings. yet.

I'm happy for any feedback, even nitpicky "that's not very zig-like" feedback. It's a new-ish language to me. Or major "this is all wrong" is fine as well.

### CODE REVIEW

**Comment by ehaas** [comment]
I know the layout engine itself is dual Apache/MIT license so that seems fine, but the test files themselves are GPLv3 licensed and are derived from the cly DSL. I'm definitely not a license expert and not entirely sure if that would be a problem with this project's MIT license. Maybe @mahkoh is around and could answer?

**Comment by TwoClocks** [comment]
Yeah. I just checked in formatted code. (I need to figure out how to get my editor to do it automatically. Or install a local githook. I always forget).

Vexu added the `__builtin_offsetof` for me a week or so ago, but I needed it to test bitfield offsets, so I just altered it for bits. Re-adding `__builtin_offsetof` is on my TODO list. I don't think anything is using it yet.

File: src/Compilation.zig
Comment (ehaas) [inline src/Compilation.zig:41]: There's an `emulate` field in `LangOpts` that was recently added that does the same thing. We could also move the `dialect` there.

**Comment by TwoClocks** [comment]
>  Maybe @mahkoh is around and could answer?

I emailed him and asked him that exact question. He pointed out that since the test aren't compiled into ARO, they are fine as GPL. Just like compiling any software w/ GCC doesn't cause that software to be released under the GPL, nor does the software being compiled affect GCC's license. I checked w/ Vexu before adding.

File: src/Parser.zig
Comment (ehaas) [inline src/Parser.zig:1746]: This can be imported at the top of the file, and for consistency should be called `RecordLayout.zig`

Comment (TwoClocks) [reply src/Parser.zig:1746]: It started as 'RecordLayout' then I happed to read the style guide in the language docs that made me change it

> File names fall into two categories: types and namespaces. If the file (implicitly a struct) has top level fields, it should be named like any other struct with fields using TitleCase. Otherwise, it should use snake_case.

File: src/record_layout.zig
Comment (ehaas) [inline src/record_layout.zig:38]: I don't think `ty` needs to be a pointer here (or in `computeLayout`). In general passing `Type` by value is preferred. In `rawAlignOf` you can recursively call it on `ty.data.sub_type.*` for the `typeof` cases. I think that function also belongs in `Type.zig`

File: src/Parser.zig
Comment (ehaas) [reply src/Parser.zig:1746]: Ah right, that makes sense. I do think it can be imported at the top of the file though.

File: src/record_layout.zig
Comment (TwoClocks) [reply src/record_layout.zig]: fixed. Yeah, that's better.
I think there is a bunch of `fn`s in `Type.zig` that take a `*const Compilation` that could just take a `Compilation`

File: src/Parser.zig
Comment (TwoClocks) [reply src/Parser.zig]: fixed

File: src/record_layout.zig
Comment (ehaas) [reply src/record_layout.zig]: Those generally all need to be pointers because `Compilation` itself is pretty heavy, and it contains allocators which usually aren't safe to copy around.

File: src/Compilation.zig
Comment (TwoClocks) [reply src/Compilation.zig]: I switched `mimic` to use `LangOpts` and deleted `dialect`.
I'll add `dialect` back in when I get there. I think I got a little ahead of myself. When I started this, I thought it was going to be a lot easier than it's turned out to be :)

File: src/Type.zig
Comment (ehaas) [inline src/Type.zig:467]: This function isn't needed; you can just use `ty.is(.@"enum")`

Comment (ehaas) [reply src/Type.zig:471]: Yes, it ultimately calls into `Type.get` which should handle typeof- and attributed types

**Review by Vexu — CHANGES_REQUESTED** [review]
I'll take a closer look at the test cases once I get Evan's enum PR merged and take a look at field align attributes.

Other things that need to be done before merging:
 * As also mentioned by Evan the `s/offsetof/bitoffsetof/` change should be reverted in favor of just adding `__builtin_bitoffsetof`
 * It would be good to add a `README.md` to `test/records` explaining where they came from and what the license situation is
 * The commits need to be cleaned up but that should only be done once the PR is otherwise ready

File: src/Compilation.zig
Comment (TwoClocks) [reply src/Compilation.zig:429]: I'm not even sure we'll need all of them. They are mostly there to make diff-ing the rust code easier. I haven't used a va_list before, but I don't think there will be any issues w/ that.
Keep the va_list in the `Record` still?

File: src/Parser.zig
Comment (TwoClocks) [reply src/Parser.zig:5534]: So in these cases that return a null-able that should never be null, would you prefer `unreachable`?

File: src/Type.zig
Comment (TwoClocks) [reply src/Type.zig:902]: yep.

File: src/record_layout.zig
Comment (TwoClocks) [reply src/record_layout.zig:309]: yes. Because of #300, I want to ignore attributes on the type, so I just take the "top" one that's on the field.

File: src/Compilation.zig
Comment (TwoClocks) [reply src/Compilation.zig:429]: My plan for `offsetof` / `bitoffset` of was to create a new `offsetof` that just called `bitoffsetof() / 8`.
But it sounds like you'd like me to revert this change and create a new `bitoffsetof`?

File: src/Parser.zig
Comment (Vexu) [reply src/Parser.zig:5534]: Yes, note that `.?` is the same as `orelse unreachable` and is preferred.

File: src/record_layout.zig
Comment (Vexu) [reply src/record_layout.zig:309]: Alright, well as long as it is changed once #300 is resolved then this is OK.

File: test/test_records.zig
Comment (Vexu) [reply test/test_records.zig:236]: No, I mean this shouldn't support `EXPECTED_ERRORS` and friends, those tests should be in `test/cases`.

File: src/Type.zig
Comment (TwoClocks) [reply src/Type.zig]: fixed.

File: src/Compilation.zig
Comment (TwoClocks) [reply src/Compilation.zig:429]: I returned `__builtin_offsetof` as it originally was, and added `__builtin_bitoffsetof`. Let me know if you'd like it done differently.

File: src/Parser.zig
Comment (TwoClocks) [reply src/Parser.zig]: fixed in a few places.

File: src/Type.zig
Comment (TwoClocks) [reply src/Type.zig]: no good reason. reverted.
None of the tests failed when I made this change though. Am I missing some?

File: src/Type.zig
Comment (TwoClocks) [reply src/Type.zig]: yep

File: test/test_records.zig
Comment (TwoClocks) [reply test/test_records.zig]: yeah. that's better. thanks.

File: test/test_records.zig
Comment (TwoClocks) [reply test/test_records.zig]: I was never calling them. deleted.

File: test/test_records.zig
Comment (TwoClocks) [reply test/test_records.zig:170]: I think I did this correctly.
If I don't call `checkExpectedErrors()` everything succeeds.

File: src/record_layout.zig
Comment (TwoClocks) [reply src/record_layout.zig]: didn't know that existed. Yep. removed.

File: src/Type.zig
Comment (Vexu) [reply src/Type.zig]: A test containing a `_Bool` bit-field wider than 1 would fail. MSVC apparently allows that based on `0036_test.c`..

**Comment by Vexu** [comment]
Just FYI, this doesn't need to be completed 100% right away, we can work to get what is currently done merged and then finish the rest in smaller steps. For starters it would be easier to add the new test harness in a separate PR.

**Comment by TwoClocks** [comment]
It's a long way from being 100% complete. Besides just working to mimic clang, and only working on x86, about 1/2 tests aren't being run. Because of things currently missing in arocc, and also because of problems I haven't figured out yet. I'm sure as you add features that help layout it will enable more test, but also likely expose more bugs in the layout code.

The test files themselves are going to evolve. I'd like to break the expected results into zig targets/compiler instead of nebulous defines. There are also some bugs in how the `EXTRA_CHECKS` are calculated in some cases.

I just did a quick test w/o the new ayout code, another 41 tests would have to be excluded, leaving only a couple that work.
At that point, it's mostly just testing the parser. 
I could alter the `test_records.zig` to just test the parser, and enable the MSVC version as well if that's what you're looking for.

File: src/Type.zig
Comment (TwoClocks) [reply src/Type.zig]: deleted after #313 merge.

File: src/Parser.zig
Comment (TwoClocks) [reply src/Parser.zig]: I'm using `hasIncompleteSize` as a proxy for "has layout happened yet?"
I think it's just paranoia though. I don't t think I need to call this as much as I do. The else branch is always 'unreachable' anyway.

File: src/record_layout.zig
Comment (TwoClocks) [reply src/record_layout.zig]: removed

File: src/record_layout.zig
Comment (TwoClocks) [reply src/record_layout.zig]: function removed after merrge of #313

**Comment by TwoClocks** [comment]
I feel like this is in a pretty good spot now.

Both Clang and Gcc work on linux-ish x86-64 targets, except for 4 tests blocked by #332 
Since adding attributes, I was able to remove all the workarounds I had in `Type.zig`.

The test files and the record test runner are cleaned up. It's set up to easily add more targets/compilers as they are implemented. The battery of tests is fairly extensive. There is a `.md` file explaining the tests, the runner, where they came from, and how to use them outside of the runner.

I feel like this is a good spot to merge, before I start working on msvc and mingw emulation.

If you could take a look and provide feedback, I'd appreciate it.

**Comment by Vexu** [comment]
Currently this is too large to be reviewed, I'd like to split this initially to a PR adding `__builtin_bitoffsetof`, this one adding the core record layout logic, and a PR just adding the new test harness and cases. I can do the bitoffsetof PR for you since it is unlikely to need any changes from you but test harness PR should probably be made by you.

**Comment by TwoClocks** [comment]
I'll put together a few small PRs for things I changed in arocc outside of `record_layout.zig`. They should probably be looked at individually.

I can split out the tests & runner as well.

**Comment by TwoClocks** [comment]
`bitoffsetof` and the tests depend on #345 being merged. I will add separate PRs for those after.

**Comment by TwoClocks** [comment]
split up into smaller PRs

**Pull request closed without merging**
